src/shared/socket-netlink.c

```diff
@@ -372,10 +332,14 @@ int in_addr_port_ifindex_name_from_string_auto(
 
         /* This accepts the following:
          * 192.168.0.1:53#example.com
-         * [2001:4860:4860::8888]:53%eth0#example.com */
-
-        /* if ret_port is NULL, then strings with port cannot be specified.
-         * Also, if ret_server_name is NULL, then server_name cannot be specified. */
+         * [2001:4860:4860::8888]:53%eth0#example.com
```

Comment:
Parsing of the ifindex in this way is ambiguous and incomplete. [RFC4007 11.7](https://tools.ietf.org/html/rfc4007#section-11.7) specifies how the zone id aka ifindex is used together with other delimiter characters. Though the RFC admits some ambiguity and "does not specify how the format for non-global addresses should be combined with the preferred format for literal IPv6 addresses", it does give a example `fe80::%2/64` and specifies that "it is important to place the zone index portion before the prefix length".

From this I conclude that the preferred way for this type of address is `[fe80::42%eth0]:53#example.com`, but only if the address is link-local and the ifindex is the "zone id" for this address. On the other hand, if the ifindex is used only to bind the socket to this interface, the preferred way is  the existing `[2001:4860:4860::8888]:53%eth0#example.com`.

So I suggest to separate these two types of ifindex, as they have different meaning in different contexts. Though in all cases both would have the same value.

Reply:
I'd be against having two syntaxes for specifying the index. We would never need both at the same time, so it seems fine to just use the second one for both purposes.